metrics/influx.go

```diff
@@ -0,0 +1,208 @@
+		Errors:         metrics.GetOrRegisterMeter(fmt.Sprintf("%s.error_count", name), r),
+		BadRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.bad_request_count", name), r),
+		JsonRequest:    metrics.GetOrRegisterMeter(fmt.Sprintf("%s.json_request_count", name), r),
+		XmlRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.xml_request_count", name), r),
+		DefinesTTL:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.defines_ttl", name), r),
+		InvalidRequest: metrics.GetOrRegisterMeter(fmt.Sprintf("%s.unknown_request_count", name), r),
+		RequestLength:  metrics.GetOrRegisterHistogram(name+".request_size_bytes", r, metrics.NewExpDecaySample(1028, 0.015)),
+	}
+}
+
+func NewInfluxConnectionMetrics(r metrics.Registry) *InfluxConnectionMetrics {
+	return &InfluxConnectionMetrics{
+		ActiveConnections:      metrics.GetOrRegisterCounter("connections.active_incoming", r),
+		ConnectionAcceptErrors: metrics.GetOrRegisterMeter("connections.accept_errors", r),
+		ConnectionCloseErrors:  metrics.GetOrRegisterMeter("connections.close_errors", r),
+	}
+}
+
+func CreateInfluxMetrics() *InfluxMetrics {
+	flushTime := time.Second * 10
+	r := metrics.NewPrefixedRegistry("prebidcache.")
+	m := &InfluxMetrics{
+		Registry:    r,
+		Puts:        NewInfluxMetricsEntry("puts.current_url", r),
+		Gets:        NewInfluxMetricsEntry("gets.current_url", r),
+		PutsBackend: NewInfluxMetricsEntryBackendPuts("puts.backend", r),
+		GetsBackend: NewInfluxMetricsEntry("gets.backend", r),
+		Connections: NewInfluxConnectionMetrics(r),
+		ExtraTTL:    &InfluxExtraTTL{ExtraTTLSeconds: metrics.GetOrRegisterHistogram("extra_ttl_seconds", r, metrics.NewUniformSample(5000))},
+	}
+
+	metrics.RegisterDebugGCStats(m.Registry)
+	metrics.RegisterRuntimeMemStats(m.Registry)
+
+	go metrics.CaptureRuntimeMemStats(m.Registry, flushTime)
+	go metrics.CaptureDebugGCStats(m.Registry, flushTime)
+
+	return m
+}
+
+// Export begins sending metrics to the configured database.
+// This method blocks indefinitely, so it should probably be run in a goroutine.
+func (m InfluxMetrics) Export(cfg config.Metrics) {
+	if cfg.Influx.Host != "" {
+		logrus.Infof("Metrics will be exported to Influx with host=%s, db=%s, username=%s", cfg.Influx.Host, cfg.Influx.Database, cfg.Influx.Username)
+		influxdb.InfluxDB(
+			m.Registry,          // metrics registry
+			time.Second*10,      // interval
+			cfg.Influx.Host,     // the InfluxDB url
+			cfg.Influx.Database, // your InfluxDB database
+			cfg.Influx.Username, // your InfluxDB user
+			cfg.Influx.Password, // your InfluxDB password
+		)
+	}
+	return
+}
+
+func (m *InfluxMetrics) RecordPutRequest(status string, duration *time.Time) {
+	if status != "" {
+		switch status {
+		case "error":
+			m.Puts.Errors.Mark(1)
+		case "bad_request":
+			m.Puts.BadRequest.Mark(1)
+		case "add":
+			m.Puts.Request.Mark(1)
+		}
+	} else if duration != nil {
+		m.Puts.Duration.UpdateSince(*duration)
```

Comment:
Why is this an `else if`? Is it not valid for both a status and a duration to be provided?

Reply:
This function gets called with one purpose or other, whether to update duration or status but not both (so far). I can change it for them to not be mutually exclusive and it wouldn't change anything in our current code. Do you prefer I remove the `else` clause?